src/api/providers/kilocode.ts

```diff
@@ -20,19 +20,62 @@ export class KiloCodeHandler extends BaseProvider implements SingleCompletionHan
 		})
 	}
 
-	async *createMessage(systemPrompt: string, messages: Anthropic.Messages.MessageParam[]): ApiStream {
+	private getIdempotencyKey(taskId: string, checkpointNumber: number): string {
+		// Create a deterministic idempotency key based on task_id and checkpoint number
+		return `${taskId}-${checkpointNumber}`
+	}
+
+	async *createMessage(
+		systemPrompt: string,
+		messages: Anthropic.Messages.MessageParam[],
+		taskId?: string,
+		checkpointNumber?: number,
+	): ApiStream {
 		let stream: AnthropicStream<Anthropic.Messages.RawMessageStreamEvent>
 		const cacheControl: CacheControlEphemeral = { type: "ephemeral" }
-		let { id: modelId, maxTokens, thinking, temperature, virtualId } = this.getModel()
+		const { id: modelId, maxTokens, thinking, temperature, virtualId } = this.getModel()
 
-		const userMsgIndices = messages.reduce(
-			(acc, msg, index) => (msg.role === "user" ? [...acc, index] : acc),
-			[] as number[],
-		)
+		// Use a for loop instead of reduce with spread to avoid linting error
+		const userMsgIndices: number[] = []
+		for (let i = 0; i < messages.length; i++) {
+			if (messages[i].role === "user") {
+				userMsgIndices.push(i)
+			}
+		}
 
 		const lastUserMsgIndex = userMsgIndices[userMsgIndices.length - 1] ?? -1
 		const secondLastMsgUserIndex = userMsgIndices[userMsgIndices.length - 2] ?? -1
 
+		// Prepare request options with headers
+		const requestOptions: { headers: Record<string, string> } = (() => {
```

Comment:
Either this typing here or the one down on 64 is redundant, making it harder to maintain.